text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,656 @@
+`Read` implementation.
+This may also be convenient elsewhere.
+
+The `io::ErrorKind` for a `ProcessError` will be:
+
+  * The `io::ErrorKind` from the spawn error, if any.
+
+  * Otherwise, a new kind `io::ErrorKind::ProcessFailed`,
+    which means that the subprocess itself failed.
+
+### Further APIs for `ProcessError`
+
+A `ProcessError` is a transparent `Default` struct so it can be
+constructed outside std, for example by async frameworks,
+or other code which handles launching subprocesses.
+
+We propose the following additional methods:
+
+```rust
+impl ProcessError {
+    /// Makes a "blank" error which doesn't contain any useful information
+    ///
+    /// `has_problem()` will return `false` until one of the setters
+    /// is used to store an actual problem.
+    //
+    // This is equivalent to `ProcessError::default()`, but provides
+    // a more semantically meaningful name, making it clear that it
+    // returns an empty error that needs filling in.
+    fn new_empty() -> Self { }
+
+    // If we keep ExitStatusError
+    fn from_exit_status_error(status: ExitStatusError) -> Self { }
+
+    /// Find out if this error contains any actual error information
+    ///
+    /// Returns `false` for a fresh blank error,
+    /// or `true` for one which has any of the error fields set.
+    /// (Currently equivalent to checking all of `status()`,
+    /// `stderr_bytes()`, `spawn_error()` and `utf8_error()`.)
+    //
+    // We must provide this because it's needed for handling programs
+    // with unusual exit status conventions (eg `diff(1)`)
+    // and a caller can't reimplement it without making assumptons
+    // about `ProcessError`'s contents.
+    fn has_problem(&self) -> bool;
+
+    /// Returns `Ok<ExitStatus>` if the only reason for the failure was a nonzero exit status.
+    /// Otherwise returns `self`.
+    ////
+    /// Use this if you want to to tolerate some exit statuses,
+    /// but still fail if there were other problems.
+    //
+    // This is optional, and could be a separate feature from the rest of the RFC.
+    // But it does make running programs like `diff` considerably easier.
+    // (It is also implementable externally in terms of .has_problem()`.)
+    pub fn just_status(self) -> Result<ExitStatus, ProcessError>;
+}
+impl Default for ProcessError { ... }
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is nontrivial new API surface.
+
+Much of the new API surface is in `ProcessError`.
+If we didn't want to try to make it easy for Rust programmers
+to run subprocesses and produce good error messages,
+we could omit this error type 
+(perhaps using something like `ExitStatusError`).
+
+Perhaps we don't need all the `read_stdout` variants,
+and could require Bob to write out the boilerplate
+or provide his own helper function.
+
+Perhaps we don't need `stdout_reader`.
+However,
+avoiding deadlocks when reading subprocess output,
+and also doing error checks properly,
+is rather subtle.
```

Comment:
Could you clarify how `stdout_reader` makes avoiding this deadlock easier?

Is this just that you need to interleave `.write` calls to the process stdin with `.read` calls to the reader, rather than calling `.write_all` and then reading all the output at once?

Reply:
The deadlocks that `stdout_reader` avoids completely are ones ike: waiting for the command's stdout EOF, and child termination, in the wrong order; waiting for stderr EOF (which can sometimes block forever for complicated reasons).

It helps only a little with deadlocks resulting from piping both into and out of the same `Command`.  (I think it may still help in those use cases, but it *doesn't* help with "you wrote too much to the command's input while it's trying to write output to you".)